src/main/java/org/reactome/addlinks/referencecreators/EnsemblBioMartMicroarrayPopulator.java

```diff
@@ -0,0 +1,159 @@
+            if (allDataStructuresPopulated(proteinToTranscripts, transcriptToProbes, uniprotToProtein)) {
+                // Retrieve Reactome Species instance and all RGPs associated with it.
+                GKInstance speciesInst = (GKInstance) adapter.fetchInstanceByAttribute(ReactomeJavaConstants.Species, ReactomeJavaConstants.name, "=", speciesName).iterator().next();
+                Collection<GKInstance> rgpInstances = adapter.fetchInstanceByAttribute(ReactomeJavaConstants.ReferenceGeneProduct, ReactomeJavaConstants.species, "=", speciesInst);
+                // Get all RGP identifiers in database, mapping them to associated instances.
+                Map<String, List<GKInstance>> rgpIdentifiersToRGPs = mapRGPIdentifiersToRGPInstances(rgpInstances);
+                // Iterate through each identifier, and then through each RGP instance.
+                for (String rgpIdentifier : rgpIdentifiersToRGPs.keySet()) {
+                    for (GKInstance rgpInst : rgpIdentifiersToRGPs.get(rgpIdentifier)) {
+                        addMicroarrayDataToReferenceGeneProduct(rgpIdentifier, rgpInst, speciesBiomartName, mappings);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Iterates through each protein associated with the RGP instance and finds transcripts associated with protein identifier.
+     * Then takes transcript identifier and finds all microarray identifiers associated with it. These microarray identifiers
+     * are then added to the 'otherIdentifier' attribute of the RGP instance. Once all microarray identifiers have been added,
+     * the instance is updated with the new data in the DB.
+     * @param rgpIdentifier - String, Identifier associated with RGP instance.
+     * @param rgpInst - GKInstance, ReferenceGeneProduct instance
+     * @param speciesBiomartName, - String, Species name in BioMart format (eg: hsapiens).
+     * @param mappings - Map<String, Map< String, List<String>>>, Super mapping structure generated by EnsemblBioMartFileProcessor.
+     * @throws Exception - Thrown by MySQLAdaptor/GKInstance methods. Typically due to accessing data that doesn't exist or from InvalidAttributeExceptions.
+     */
+    private void addMicroarrayDataToReferenceGeneProduct(String rgpIdentifier, GKInstance rgpInst, String speciesBiomartName, Map<String, Map<String, List<String>>> mappings) throws Exception {
+        // Retrieve protein identifier associated with reference DB.
+        Set<String> rgpProteins = findRGPProteins(rgpIdentifier, rgpInst, speciesBiomartName, mappings);
+        // For each protein identifier retrieved, find microarray identifiers associated and
+        // add them to the 'otherIdentifier' attribute of the RGP instance.
+        for (String protein : rgpProteins) {
+            if (EnsemblBioMartUtil.getProteinToTranscriptsMappings(speciesBiomartName, mappings).containsKey(protein)) {
```

Comment:
The conditions in the for loop can be removed if the `EnsemblBioMartUtil` class methods returning the mappings are guaranteed to return a map object (either populated or empty) instead of a null value.

This can be done by using the `computeIfAbsent` method in place of the `get` method in the `EnsemblBioMartUtil` mappings getter methods.  For example:

```
public static Map<String, List<String>> getProteinToTranscriptsMappings(String speciesBiomartName, Map<String, Map<String, List<String>>> mappings) {
    String proteinToTranscriptsKey = speciesBiomartName + PROTEIN_TO_TRANSCRIPTS_SUFFIX;
    return mappings.computeIfAbsent(proteinToTranscriptsKey, k -> new HashMap<>());
}
```

Reply:
Unfortunately, I am still getting NullPointerExceptions after changing the method to use `computeIfAbsent`. It works fine with the inner check, `getTranscriptToMicroarrays` (T2M), but not for the outer one, `getProteinToTranscriptMappings` (P2T). I think for the outer one that it must be trying to access Protein values (using RGP identifiers) that do not exist in the maps. For the inner one, it is trying to access a Transcript value in TranscriptToMicroarrays (T2M) from the P2T structure, which would always exist since they both come from the same line in the same file.

That might be a very wordy explanation, so to summarize: 
- The outer check (P2T) is accessed using an [protein] identifier from an RGP, which may or may not have a key in the mapping.
- The inner check (T2M) is accessed using a transcript identifier from P2T, and since these values come from the same file will always exist. 

I think `computeIfAbsent` doesn't make a difference here since at the end of the day you are _sometimes_  trying to access a key that does not exist in P2T. I changed all getters to use `computeIfAbsent` though in EnsemblBioMartUtils, and removed the inner check.

Reply:
Ah, I see.  I think the `get` methods used here with the `for` loops would also need to use `computeIfAbsent` so that an empty array could be returned in the even the key doesn't exist in the P2T map.